### PR TITLE
CLN: make Info and DataFrameInfo subclasses

### PR DESCRIPTION
**Pull request by MarcoGorelli**
precursor to #31796

Makes Info class and DataFrameInfo subclass so there's no need for all the
```python
if isinstance(data, ABCDataFrame)
```

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -2460,11 +2460,11 @@ def to_html(
             <class 'pandas.core.frame.DataFrame'>
             RangeIndex: 5 entries, 0 to 4
             Data columns (total 3 columns):
-                #   Column     Non-Null Count  Dtype
+             #   Column     Non-Null Count  Dtype
             ---  ------     --------------  -----
-                0   int_col    5 non-null      int64
-                1   text_col   5 non-null      object
-                2   float_col  5 non-null      float64
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:2467]: I'd accidentally introduced these extra whitespaces when I indented the docstrings as part of a previous PR, and so am fixing them now

File: pandas/io/formats/info.py [hunk 2]
```diff
@@ -39,115 +40,208 @@ def _put_str(s: Union[str, Dtype], space: int) -> str:
...
+            memory_usage = get_option("display.memory_usage")
+
+        self.data = data
+        self.verbose = verbose
+        self.buf = buf
+        self.max_cols = max_cols
+        self.memory_usage = memory_usage
+        self.null_counts = null_counts
+
+    def _get_mem_usage(self, deep: bool) -> int:
```
Comment (WillAyd) [inline pandas/io/formats/info.py:65]: I think easier if you just make this an ABC

File: pandas/io/formats/info.py [hunk 3]
```diff
@@ -39,115 +40,208 @@ def _put_str(s: Union[str, Dtype], space: int) -> str:
...
-    max_rows = get_option("display.max_info_rows", len(data) + 1)
-
-    if null_counts is None:
-        show_counts = (col_count <= max_cols) and (len(data) < max_rows)
-    else:
-        show_counts = null_counts
-    exceeds_info_cols = col_count > max_cols
-
-    def _verbose_repr():
+class Info:
```
Comment (WillAyd) [inline pandas/io/formats/info.py:43]: Can you name this `BaseInfo`?

File: pandas/io/formats/info.py [hunk 4]
```diff
@@ -39,115 +41,212 @@ def _put_str(s: Union[str, Dtype], space: int) -> str:
...
+    def __init__(
+        self,
+        data: FrameOrSeries,
+        verbose: Optional[bool] = None,
+        buf: Optional[IO[str]] = None,
+        max_cols: Optional[int] = None,
+        memory_usage: Optional[Union[bool, str]] = None,
+        null_counts: Optional[bool] = None,
+    ):
+        if buf is None:  # pragma: no cover
```
Comment (jreback) [inline pandas/io/formats/info.py:54]: why do you need the pragma here?

File: pandas/io/formats/info.py [hunk 5]
```diff
@@ -39,115 +41,212 @@ def _put_str(s: Union[str, Dtype], space: int) -> str:
...
+
+        max_rows = get_option("display.max_info_rows", len(self.data) + 1)
+
+        if self.null_counts is None:
+            show_counts = (col_count <= max_cols) and (len(self.data) < max_rows)
+        else:
+            show_counts = self.null_counts
+        exceeds_info_cols = col_count > max_cols
+
+        def _sizeof_fmt(num, size_qualifier):
```
Comment (jreback) [inline pandas/io/formats/info.py:160]: move this to a module level function

File: pandas/io/formats/info.py [hunk 6]
```diff
@@ -39,115 +41,212 @@ def _put_str(s: Union[str, Dtype], space: int) -> str:
...
+        Get DataFrame's memory usage in bytes.
+
+        Parameters
+        ----------
+        deep : bool
+            If True, introspect the data deeply by interrogating object dtypes
+            for system-level memory consumption, and include it in the returned
+            values.
+
+        Returns
```
Comment (jreback) [inline pandas/io/formats/info.py:212]: you can put all of the doc-strings on the base class

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good

File: pandas/io/formats/info.py (hunk 4 above)
Comment (WillAyd) [reply pandas/io/formats/info.py:54]: Noticed this is in the current implementation. Not sure of reasoning but yea probably safe to remove

Comment (MarcoGorelli) [reply pandas/io/formats/info.py:85]: Yes, it was from the current implementation - have removed

**Comment by jreback** [comment]
looks good can you merge master and ping on green.

**Comment by MarcoGorelli** [comment]
> looks good can you merge master and ping on green.

Thanks @jreback - green now

**Comment by MarcoGorelli** [comment]
Friendly ping :smile: (sorry to do this, but I can submit a PR for `Series.info` if/when this is merged)

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli very nice

**Pull request merged**